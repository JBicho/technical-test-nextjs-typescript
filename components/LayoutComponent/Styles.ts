import styled, { keyframes } from 'styled-components';

const gradientAnimation = keyframes`
  0% { background-position: 0% 50%; }
  50% { background-position: 100% 50%; }
  100% { background-position: 0% 50%; }
`;

const Background = styled.div`
  background: linear-gradient(135deg, #1a1a1a, #6e8ef9, #9b51e0, #000000);
  background-size: 200% 200%;
  animation: ${gradientAnimation} 10s ease infinite;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
`;

const OverlayPattern = styled.div`
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-image: url('https://www.transparenttextures.com/patterns/white-diamond.png');
  opacity: 0.05;
  pointer-events: none;
`;

const Menu = styled.nav`
  position: fixed;
  background: rgba(110, 142, 249, 1);
  box-shadow: var(--global-box-shadow);
  width: 100%;
  top: 0;
  padding: var(--global-rhythm);
  z-index: 9999;

  a {
    cursor: pointer;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;

    span {
      display: flex;
      align-items: center;
      margin-left: calc(var(--global-rhythm) / 2);
    }
  }
`;

const Container = styled.div`
  padding: 0 var(--global-rhythm);
`;

const Main = styled.div`
  padding: calc(2 * var(--global-rhythm)) 0;
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 100vh;

  h1 {
    margin-top: 90px;
  }
`;

export { Background, Container, Main, Menu, OverlayPattern };
