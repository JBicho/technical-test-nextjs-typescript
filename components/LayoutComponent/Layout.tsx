import type { ReactElement } from 'react';
import { Background, Container, Main, Menu, OverlayPattern } from './Styles';
import Image from 'next/image';
import { ConditionalLink } from '../ConditionalLink/ConditionalLink';

export const Layout = (page: ReactElement): JSX.Element => {
  return (
    <>
      <Background>
        <OverlayPattern />
        <Menu>
          <ConditionalLink href={'/'} aria-label="Go to home page">
            <Image
              priority={true}
              src="/images/pokeball.png"
              alt="Pokeball Image"
              width={60}
              height={60}
            />
            <span>Home</span>
          </ConditionalLink>
        </Menu>
        <Container>
          <Main>{page}</Main>
        </Container>
      </Background>
    </>
  );
};
