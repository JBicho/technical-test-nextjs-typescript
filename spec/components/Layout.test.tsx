import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import type { ReactElement } from 'react';
import { Layout } from '../../components/LayoutComponent/Layout';
import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));
jest.mock('next/image', () => ({
  __esModule: true,
  default: ({ src, alt, width, height }: any) => (
    <img src={src} alt={alt} width={width} height={height} />
  ),
}));

describe('Layout Component', () => {
  const mockRouter = {
    pathname: '/some-page',
    push: jest.fn(),
  };

  const mockPage: ReactElement = <div>Mock Page Content</div>;

  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  const setupComponent = () => {
    return render(<>{Layout(mockPage)}</>);
  };

  it('Should match snapshot with a mock page', () => {
    const { asFragment } = setupComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should match snapshot when on homepage', () => {
    (useRouter as jest.Mock).mockReturnValue({
      ...mockRouter,
      pathname: '/',
    });

    const { asFragment } = setupComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the page content', () => {
    const { getByText } = setupComponent();
    const content = getByText('Mock Page Content');

    expect(content).toBeInTheDocument();
  });

  it('Should render the home link and image', () => {
    const { getByAltText, getByText } = setupComponent();
    const image = getByAltText('Pokeball Image');
    const link = getByText('Home');

    expect(image).toBeInTheDocument();
    expect(link).toBeInTheDocument();
  });

  it('Should render with proper attributes', () => {
    const { getByAltText, getByRole } = setupComponent();
    const image = getByAltText('Pokeball Image');
    const link = getByRole('link', { name: /home/i });

    expect(image).toHaveAttribute('width', '60');
    expect(image).toHaveAttribute('height', '60');
    expect(link).toHaveAttribute('href', '/');
  });
});
