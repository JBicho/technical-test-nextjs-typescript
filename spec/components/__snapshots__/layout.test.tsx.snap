// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Component Should match snapshot when on homepage 1`] = `
<DocumentFragment>
  <div
    class="Styles__Background-sc-1o8hx7p-0 dKNCZO"
  >
    <div
      class="Styles__OverlayPattern-sc-1o8hx7p-1 kEikSX"
    />
    <nav
      class="Styles__Menu-sc-1o8hx7p-2 iABhsv"
    >
      <a
        href="#"
      >
        <span
          aria-disabled="true"
        >
          <img
            alt="Pokeball Image"
            height="60"
            src="/images/pokeball.png"
            width="60"
          />
          <span>
            Home
          </span>
        </span>
      </a>
    </nav>
    <div
      class="Styles__Container-sc-1o8hx7p-3 iOvVmP"
    >
      <div
        class="Styles__Main-sc-1o8hx7p-4 hDokoW"
      >
        <div>
          Mock Page Content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Layout Component Should match snapshot with a mock page 1`] = `
<DocumentFragment>
  <div
    class="Styles__Background-sc-1o8hx7p-0 dKNCZO"
  >
    <div
      class="Styles__OverlayPattern-sc-1o8hx7p-1 kEikSX"
    />
    <nav
      class="Styles__Menu-sc-1o8hx7p-2 iABhsv"
    >
      <a
        href="/"
      >
        <span
          aria-disabled="false"
        >
          <img
            alt="Pokeball Image"
            height="60"
            src="/images/pokeball.png"
            width="60"
          />
          <span>
            Home
          </span>
        </span>
      </a>
    </nav>
    <div
      class="Styles__Container-sc-1o8hx7p-3 iOvVmP"
    >
      <div
        class="Styles__Main-sc-1o8hx7p-4 hDokoW"
      >
        <div>
          Mock Page Content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
